keyless/pepper/service/src/groth16_vk.rs

```diff
@@ -0,0 +1,120 @@
+            gamma_g2: g2_from_api_repr(&self.data.gamma_g2)
+                .map_err(|e| anyhow!("to_ark_pvk() failed with gamma_g2 convert err: {e}"))?,
+            delta_g2: g2_from_api_repr(&self.data.delta_g2)
+                .map_err(|e| anyhow!("to_ark_pvk() failed with delta_g2 convert err: {e}"))?,
+            gamma_abc_g1,
+        };
+        Ok(PreparedVerifyingKey::from(ark_vk))
+    }
+}
+
+/// This variable holds the cached on-chain VK. A refresh loop exists to update it periodically.
+pub static ONCHAIN_GROTH16_VK: Lazy<Arc<RwLock<Option<OnChainGroth16VerificationKey>>>> =
+    Lazy::new(|| Arc::new(RwLock::new(None)));
+
+fn g1_from_api_repr(api_repr: &str) -> Result<G1Affine> {
+    let bytes = unhexlify_api_bytes(api_repr)
+        .map_err(|e| anyhow!("g1_from_api_repr() failed with unhex err: {e}"))?;
+    let ret = G1Affine::deserialize_compressed(bytes.as_slice())
+        .map_err(|e| anyhow!("g1_from_api_repr() failed with g1 deser err: {e}"))?;
+    Ok(ret)
+}
+
+fn g2_from_api_repr(api_repr: &str) -> Result<G2Affine> {
+    let bytes = unhexlify_api_bytes(api_repr)
+        .map_err(|e| anyhow!("g2_from_api_repr() failed with unhex err: {e}"))?;
+    let ret = G2Affine::deserialize_compressed(bytes.as_slice())
+        .map_err(|e| anyhow!("g2_from_api_repr() failed with g2 deser err: {e}"))?;
+    Ok(ret)
+}
+
+impl ExternalResource for OnChainGroth16VerificationKey {
+    fn resource_name() -> String {
+        "OnChainGroth16VerificationKey".to_string()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::groth16_vk::{OnChainGroth16VerificationKey, VKeyData};
+    use ark_bn254::{Bn254, Fr, G1Affine, G2Affine};
+    use ark_groth16::Groth16;
+    use ark_serialize::CanonicalDeserialize;
+
+    #[test]
+    fn test_to_ark_pvk() {
+        let api_repr = OnChainGroth16VerificationKey {
+            r#type: "0x1::keyless_account::Groth16VerificationKey".to_string(),
+            data: VKeyData {
+                alpha_g1: "0xe39cb24154872dbdbbdbc8056c6eb3e6cab3ad82f80ded72ed4c9301c5b3da15".to_string(),
+                beta_g2: "0x9a732e38644f89ad2c7bd629b84d6b81f2e83ca4b3cddfd99c0254e49332861e2fcec4f74545abdd42c8857ff8df6d3f6b3670f930d1d5ba961655ea38ded315".to_string(),
+                delta_g2: "0x04c7b3a2734731369a281424c2bd7af229b92496527fd0a01bfe4a5c01e0a92f256921817b6d6cf040ccd483d81738ac88571b57009f182946e8a88cced03a01".to_string(),
+                gamma_abc_g1: vec![
+                    "0x2f4f4bc4acbea0c3bae9e676fb59537e2e46994d5896e286e6fcccc7e14b1b2d".to_string(),
+                    "0x979308443fbac05f6d22a16525c26246e965a9be68e163154f44b20d6b2ddf18".to_string(),
+                ],
+                gamma_g2: "0xedf692d95cbdde46ddda5ef7d422436779445c5e66006a42761e1f12efde0018c212f3aeb785e49712e7a9353349aaf1255dfb31b7bf60723a480d9293938e19".to_string(),
+            },
+        };
+
+        let ark_pvk = api_repr.to_ark_pvk().unwrap();
```

Comment:
As always, hard-coding proofs in test-cases without writing instructions for how to generate them is a big no no...

Reply:
This is a bigger problem throughout our repo, which I hope we can address eventually.

Reply:
i can add the source.